File: benchmarks/compute.c
#include <stdio.h>
#include <stdint.h>
#include <time.h>

int64_t collatz_len(int64_t n) {
    int64_t len = 0;
    while (n > 1) {
        if (n % 2 == 0) {
            n = n / 2;
        } else {
            n = n * 3 + 1;
        }
        len++;
    }
    return len;
}

int64_t longest_collatz(int64_t limit) {
    int64_t best = 0;
    for (int64_t i = 1; i < limit; i++) {
        int64_t len = collatz_len(i);
        if (len > best) {
            best = len;
        }
    }
    return best;
}

int64_t is_prime(int64_t n) {
    if (n < 2) return 0;
    if (n == 2) return 1;
    if (n % 2 == 0) return 0;
    for (int64_t d = 3; d * d <= n; d += 2) {
        if (n % d == 0) return 0;
    }
    return 1;
}

int64_t count_primes(int64_t limit) {
    int64_t count = 0;
    for (int64_t i = 2; i < limit; i++) {
        if (is_prime(i)) count++;
    }
    return count;
}

int64_t ack(int64_t m, int64_t n) {
    if (m == 0) return n + 1;
    if (n == 0) return ack(m - 1, 1);
    return ack(m - 1, ack(m, n - 1));
}

int64_t popcount(int64_t n) {
    int64_t count = 0;
    while (n > 0) {
        count += n & 1;
        n >>= 1;
    }
    return count;
}

int64_t total_popcount(int64_t limit) {
    int64_t total = 0;
    for (int64_t i = 1; i <= limit; i++) {
        total += popcount(i);
    }
    return total;
}

int64_t tak(int64_t x, int64_t y, int64_t z) {
    if (x <= y) return z;
    return tak(tak(x - 1, y, z), tak(y - 1, z, x), tak(z - 1, x, y));
}

int64_t gcd(int64_t a, int64_t b) {
    while (b != 0) {
        int64_t t = a % b;
        a = b;
        b = t;
    }
    return a;
}

int64_t euler_totient(int64_t n) {
    int64_t count = 0;
    for (int64_t i = 1; i < n; i++) {
        if (gcd(i, n) == 1) count++;
    }
    return count;
}

int64_t sum_totients(int64_t limit) {
    int64_t total = 0;
    for (int64_t i = 1; i <= limit; i++) {
        total += euler_totient(i);
    }
    return total;
}

int64_t digit_sum(int64_t n) {
    int64_t sum = 0;
    while (n > 0) {
        sum += n % 10;
        n /= 10;
    }
    return sum;
}

int64_t total_digit_sum(int64_t limit) {
    int64_t total = 0;
    for (int64_t i = 1; i <= limit; i++) {
        total += digit_sum(i);
    }
    return total;
}

int64_t get_time_ns() {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (int64_t)ts.tv_sec * 1000000000LL + (int64_t)ts.tv_nsec;
}

int main() {
    printf("=== C Compute Benchmarks ===\n");

    // Benchmark 1: Collatz conjecture
    int64_t limit = 1000000;
    int64_t start = get_time_ns();
    int64_t result = longest_collatz(limit);
    int64_t elapsed = get_time_ns() - start;
    printf("Collatz longest chain under %ld:\n", limit);
    printf("  Time: %ld ms\n", elapsed / 1000000);
    printf("  Result: %ld\n", result);

    // Benchmark 2: Prime counting
    limit = 1000000;
    start = get_time_ns();
    result = count_primes(limit);
    elapsed = get_time_ns() - start;
    printf("Prime count under %ld:\n", limit);
    printf("  Time: %ld ms\n", elapsed / 1000000);
    printf("  Result: %ld\n", result);

    // Benchmark 3: Ackermann
    int64_t m = 3, n = 11;
    start = get_time_ns();
    result = ack(m, n);
    elapsed = get_time_ns() - start;
    printf("Ackermann(%ld, %ld):\n", m, n);
    printf("  Time: %ld ms\n", elapsed / 1000000);
    printf("  Result: %ld\n", result);

    // Benchmark 4: Popcount
    limit = 10000000;
    start = get_time_ns();
    result = total_popcount(limit);
    elapsed = get_time_ns() - start;
    printf("Popcount sum 1..%ld:\n", limit);
    printf("  Time: %ld ms\n", elapsed / 1000000);
    printf("  Result: %ld\n", result);

    // Benchmark 5: Takeuchi
    int64_t x = 24, y = 16, z = 8;
    start = get_time_ns();
    result = tak(x, y, z);
    elapsed = get_time_ns() - start;
    printf("Tak(%ld, %ld, %ld):\n", x, y, z);
    printf("  Time: %ld ms\n", elapsed / 1000000);
    printf("  Result: %ld\n", result);

    // Benchmark 6: Euler totient sum
    limit = 10000;
    start = get_time_ns();
    result = sum_totients(limit);
    elapsed = get_time_ns() - start;
    printf("Euler totient sum(1..%ld):\n", limit);
    printf("  Time: %ld ms\n", elapsed / 1000000);
    printf("  Result: %ld\n", result);

    // Benchmark 7: Digit sum
    limit = 10000000;
    start = get_time_ns();
    result = total_digit_sum(limit);
    elapsed = get_time_ns() - start;
    printf("Digit sum(1..%ld):\n", limit);
    printf("  Time: %ld ms\n", elapsed / 1000000);
    printf("  Result: %ld\n", result);

    return 0;
}

File: benchmarks/compute.rs
use std::hint::black_box;
use std::time::Instant;

const RUNS: usize = 3;

fn collatz_len(mut n: i64) -> i64 {
    let mut length: i64 = 0;
    while n > 1 {
        if n % 2 == 0 {
            n /= 2;
        } else {
            n = n * 3 + 1;
        }
        length += 1;
    }
    length
}

fn longest_collatz(limit: i64) -> i64 {
    let mut best: i64 = 0;
    for i in 1..limit {
        let length = collatz_len(i);
        if length > best {
            best = length;
        }
    }
    best
}

fn is_prime(n: i64) -> bool {
    if n < 2 { return false; }
    if n == 2 { return true; }
    if n % 2 == 0 { return false; }
    let mut d: i64 = 3;
    while d * d <= n {
        if n % d == 0 { return false; }
        d += 2;
    }
    true
}

fn count_primes(limit: i64) -> i64 {
    let mut count: i64 = 0;
    for i in 2..limit {
        if is_prime(i) { count += 1; }
    }
    count
}

fn ack(m: i64, n: i64) -> i64 {
    if m == 0 { return n + 1; }
    if n == 0 { return ack(m - 1, 1); }
    ack(m - 1, ack(m, n - 1))
}

fn popcount(mut n: i64) -> i64 {
    let mut count: i64 = 0;
    while n > 0 {
        count += n & 1;
        n >>= 1;
    }
    count
}

fn total_popcount(limit: i64) -> i64 {
    let mut total: i64 = 0;
    for i in 1..=limit {
        total += popcount(i);
    }
    total
}

fn tak(x: i64, y: i64, z: i64) -> i64 {
    if x <= y { return z; }
    tak(tak(x - 1, y, z), tak(y - 1, z, x), tak(z - 1, x, y))
}

fn gcd(mut a: i64, mut b: i64) -> i64 {
    while b != 0 {
        let t = a % b;
        a = b;
        b = t;
    }
    a
}

fn euler_totient(n: i64) -> i64 {
    let mut count: i64 = 0;
    for i in 1..n {
        if gcd(i, n) == 1 { count += 1; }
    }
    count
}

fn sum_totients(limit: i64) -> i64 {
    let mut total: i64 = 0;
    for i in 1..=limit {
        total += euler_totient(i);
    }
    total
}

fn digit_sum(mut n: i64) -> i64 {
    let mut sum: i64 = 0;
    while n > 0 {
        sum += n % 10;
        n /= 10;
    }
    sum
}

fn total_digit_sum(limit: i64) -> i64 {
    let mut total: i64 = 0;
    for i in 1..=limit {
        total += digit_sum(i);
    }
    total
}

fn main() {
    println!("=== Rust Compute Benchmarks ===");

    // Benchmark 1: Collatz conjecture
    let limit: i64 = 1000000;
    longest_collatz(black_box(limit)); // warmup
    let mut best: u128 = u128::MAX;
    let mut result: i64 = 0;
    for _ in 0..RUNS {
        let start = Instant::now();
        let r = longest_collatz(black_box(limit));
        let elapsed = start.elapsed().as_nanos();
        if elapsed < best { best = elapsed; result = r; }
    }
    println!("Collatz longest chain under {}:", limit);
    println!("  Time: {} ms", best / 1000000);
    println!("  Result: {}", result);

    // Benchmark 2: Prime counting
    let limit: i64 = 1000000;
    count_primes(black_box(limit)); // warmup
    best = u128::MAX;
    for _ in 0..RUNS {
        let start = Instant::now();
        let r = count_primes(black_box(limit));
        let elapsed = start.elapsed().as_nanos();
        if elapsed < best { best = elapsed; result = r; }
    }
    println!("Prime count under {}:", limit);
    println!("  Time: {} ms", best / 1000000);
    println!("  Result: {}", result);

    // Benchmark 3: Ackermann
    let (m, n) = (3i64, 11i64);
    ack(black_box(m), black_box(n)); // warmup
    best = u128::MAX;
    for _ in 0..RUNS {
        let start = Instant::now();
        let r = ack(black_box(m), black_box(n));
        let elapsed = start.elapsed().as_nanos();
        if elapsed < best { best = elapsed; result = r; }
    }
    println!("Ackermann({}, {}):", m, n);
    println!("  Time: {} ms", best / 1000000);
    println!("  Result: {}", result);

    // Benchmark 4: Popcount
    let limit: i64 = 10000000;
    total_popcount(black_box(limit)); // warmup
    best = u128::MAX;
    for _ in 0..RUNS {
        let start = Instant::now();
        let r = total_popcount(black_box(limit));
        let elapsed = start.elapsed().as_nanos();
        if elapsed < best { best = elapsed; result = r; }
    }
    println!("Popcount sum 1..{}:", limit);
    println!("  Time: {} ms", best / 1000000);
    println!("  Result: {}", result);

    // Benchmark 5: Takeuchi
    let (x, y, z) = (24i64, 16i64, 8i64);
    tak(black_box(x), black_box(y), black_box(z)); // warmup
    best = u128::MAX;
    for _ in 0..RUNS {
        let start = Instant::now();
        let r = tak(black_box(x), black_box(y), black_box(z));
        let elapsed = start.elapsed().as_nanos();
        if elapsed < best { best = elapsed; result = r; }
    }
    println!("Tak({}, {}, {}):", x, y, z);
    println!("  Time: {} ms", best / 1000000);
    println!("  Result: {}", result);

    // Benchmark 6: Euler totient sum
    let limit: i64 = 10000;
    sum_totients(black_box(limit)); // warmup
    best = u128::MAX;
    for _ in 0..RUNS {
        let start = Instant::now();
        let r = sum_totients(black_box(limit));
        let elapsed = start.elapsed().as_nanos();
        if elapsed < best { best = elapsed; result = r; }
    }
    println!("Euler totient sum(1..{}):", limit);
    println!("  Time: {} ms", best / 1000000);
    println!("  Result: {}", result);

    // Benchmark 7: Digit sum
    let limit: i64 = 10000000;
    total_digit_sum(black_box(limit)); // warmup
    best = u128::MAX;
    for _ in 0..RUNS {
        let start = Instant::now();
        let r = total_digit_sum(black_box(limit));
        let elapsed = start.elapsed().as_nanos();
        if elapsed < best { best = elapsed; result = r; }
    }
    println!("Digit sum(1..{}):", limit);
    println!("  Time: {} ms", best / 1000000);
    println!("  Result: {}", result);
}
